semeio/workflows/localisation/local_script_lib.py

```diff
@@ -207,34 +210,327 @@ def activate_gen_param(
+        has_region = region_parameter == region_number
+        if has_region.any():
+            regions_in_param.append(region_number)
+
+    # Find all region parameter values matching any of the regions defined
+    # to be used in localisation
+    active_region_values_used = np.zeros(len(region_parameter), dtype=np.int32)
+    active_region_values_used[:] = missing_code
+    for region_number in active_segment_list:
+        found_values = region_parameter == region_number
+        active_region_values_used[found_values] = region_number
+
+    # Calculate scaling values using the user config input
+    index_per_used_region = np.zeros((max_region_number + 1), dtype=np.uint16)
+    index_per_used_region[:] = missing_code
+
+    # Both  active_segment_list and scaling_value_list are equal in length.
+    # The region number active_segment_list[index] is specified
+    # to have scaling factor scaling_value_list[index].
+    # Convert active_segment_list and scaling_value_list to numpy array
+    active_segment_array = np.array(active_segment_list)
+    scaling_value_array = np.array(scaling_value_list)
+
+    index_values = np.arange(len(active_segment_list))
+    index_per_used_region[active_segment_array[index_values]] = index_values
+
+    # Select all defined region values and omit all region values not
+    # specified to be active by the user as well as all missing values
+    # originally within the region_parameter.
+    use_scaling_regions = active_region_values_used != missing_code
+    selected_region_values = active_region_values_used[use_scaling_regions]
+
+    # Calculate scaling values
+    scaling_values_active = scaling_value_array[
+        index_per_used_region[selected_region_values]
+    ]
+
+    # Create a full sized 3D parameter where all but the specified active
+    # regions have 0 scaling value.
+    scaling_values = np.zeros(len(region_parameter), dtype=np.float32)
+    scaling_values[use_scaling_regions] = scaling_values_active
+
+    return scaling_values, active_region_values_used, regions_in_param
+
+
+def smooth_parameter(
+    grid, smooth_range_list, scaling_values, active_localisation_region, missing_code
+):
+    nx = grid.get_nx()
+    ny = grid.get_ny()
+    nz = grid.get_nz()
+    di = smooth_range_list[0]
+    dj = smooth_range_list[1]
+    scaling_values_smooth = np.zeros(nx * ny * nz, dtype=np.float32)
+    for k, j0, i0 in itertools.product(range(nz), range(ny), range(nx)):
+        index0 = i0 + j0 * nx + k * nx * ny
+        if active_localisation_region[index0] != missing_code:
+            sumv = 0.0
+            nval = 0
+            ilow = max(0, i0 - di)
+            ihigh = min(i0 + di + 1, nx)
+            jlow = max(0, j0 - dj)
+            jhigh = min(j0 + dj + 1, ny)
+            for i in range(ilow, ihigh):
+                for j in range(jlow, jhigh):
+                    index = i + j * nx + k * nx * ny
+                    if active_localisation_region[index] != missing_code:
```

Comment:
Not sure, maybe? The important thing here is that the average calculated in this simple smoothing algorithm is around each active grid cell that also are within active regions and that inactive grid cells as well as grid cells belonging to inactive regions is not part of the averaging. The two if statements makes this clear since the input has marked all cell values not to be used by a missing code.